### PR TITLE
HPCC-15660 Can't pass stored values to EMBED options in Thor

### PR DESCRIPTION
**Pull request by richardkchapman**
Rework the code generated for EMBED options so that it is calculated outside
of the generated user function and passed in. This will mean that it can be
calculated just once (on the master, in Thor) and serialized to the slaves.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review. Could consider retargetting to 6.0.2

File: ecl/hqlcpp/hqlcpp.cpp [hunk 1]
```diff
@@ -11767,40 +11767,19 @@ void HqlCppTranslator::buildScriptFunctionDefinition(BuildCtx &funcctx, IHqlExpr
...
-
-            IHqlExpression * value = child->queryChild(0);
-            if (value)
-                attrParam.append("=");
-            attrArgs.append(*createConstant(attrParam));
-            if (value)
-                attrArgs.append(*ensureExprType(value, unknownStringType));
-        }
+        optionsParam = formals->queryChild(formals->numChildren()-1);
+        assertex(streq(str(optionsParam->queryId()), "__options"));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp]: Normally __optionsId would be added to hqlcatom.?pp, this test would become a pointer comparison, and there would be no createIdAtom() call below.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 2]
```diff
@@ -11821,73 +11800,76 @@ void HqlCppTranslator::buildScriptFunctionDefinition(BuildCtx &funcctx, IHqlExpr
...
         IHqlExpression * param = formals->queryChild(i);
-        ITypeInfo *paramType = param->queryType();
-        IIdAtom * paramId = param->queryId();
-        const char * paramNameText = str(paramId);
-        if (!options.preserveCaseExternalParameter)
-            paramNameText = str(lower(paramId));
-        args.append(*createConstant(paramNameText));
-        IIdAtom * bindFunc;
-        switch (paramType->getTypeCode())
+        if (param != optionsParam)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp]: minor, could be moved earlier.  Possibly simpler as if (param == optionParam) continue;

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 3]
```diff
@@ -6224,18 +6224,81 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+            if (cast != boundCast)
+                throwError(HQLERR_EmbedParamNotSupportedInOptions);
+
+            folded.setown(foldHqlExpression(cast));
+        }
+        else
+            folded.setown(createConstant(""));
+        HqlExprArray newFormals;
+        HqlExprArray attrs;
+        unwindChildren(newFormals, formals, 0);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: trivial: No need to specify ,0 on the end.

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 4]
```diff
@@ -6224,18 +6224,81 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+        if (defaults)
+            unwindChildren(newDefaults, defaults, 0);
+        while (newDefaults.length() < formals->numChildren())
+            newDefaults.append(*createOmittedValue());
+        newDefaults.append(*folded.getClear());
+        formals = formals->clone(newFormals);
+        defaults = createValueSafe(no_sortlist, makeSortListType(NULL), newDefaults);
+        funcdefArgs.append(*LINK(formals));
+        funcdefArgs.append(*LINK(defaults));
+        unwindChildren(funcdefArgs, newFuncDef, 3);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: I suspect 3 should be 2

Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: ignore that.

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 5]
```diff
@@ -6224,18 +6224,81 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+
+        HqlExprArray newDefaults;
+        if (defaults)
+            unwindChildren(newDefaults, defaults, 0);
+        while (newDefaults.length() < formals->numChildren())
+            newDefaults.append(*createOmittedValue());
+        newDefaults.append(*folded.getClear());
+        formals = formals->clone(newFormals);
+        defaults = createValueSafe(no_sortlist, makeSortListType(NULL), newDefaults);
+        funcdefArgs.append(*LINK(formals));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: may need some more calls to inheritDependencies - e.g., on all the expressions used to build up newDefaults.  Test by having a complex default value for the stored value used as a parameter.

Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: That may start getting complicated.  I suspect this transformation should actually be done in HqlTreeNormalizer instead.

File: ecl/hqlcpp/hqlcpp.cpp (hunk 2 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlcpp.cpp]: Changed

File: ecl/hqlcpp/hqlttcpp.cpp (hunk 3 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlttcpp.cpp]: Changed

File: ecl/hqlcpp/hqlcpp.cpp (hunk 1 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlcpp.cpp]: Changed

File: ecl/hqlcpp/hqlttcpp.cpp (hunk 5 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlttcpp.cpp]: I tried and failed to reproduce an issue in this area...

**Comment by richardkchapman** [comment]
Closing in favour of 6.0.2 version

**Pull request closed without merging**